src/Symfony/Component/Serializer/NameConverter/MetadataAwareNameConverter.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\NameConverter;
+
+use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactoryInterface;
+
+/**
+ * @author Fabien Bourigault <bourigaultfabien@gmail.com>
+ */
+final class MetadataAwareNameConverter implements AdvancedNameConverterInterface
+{
+    /**
+     * @var ClassMetadataFactoryInterface
+     */
+    private $metadataFactory;
+
+    public function __construct(ClassMetadataFactoryInterface $metadataFactory)
+    {
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($propertyName, string $class = null, string $format = null, array $context = array())
+    {
+        if (null === $class || !$this->metadataFactory->hasMetadataFor($class)) {
+            return $propertyName;
+        }
+
+        $classMetadata = $this->metadataFactory->getMetadataFor($class);
+        $attributesMetadata = $classMetadata->getAttributesMetadata();
+
+        if (!isset($attributesMetadata[$propertyName])) {
+            return $propertyName;
+        }
+
+        $metadata = $attributesMetadata[$propertyName];
+
+        if (null === $metadata->getSerializedName()) {
+            return $propertyName;
+        }
+
+        return $metadata->getSerializedName();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function denormalize($propertyName, string $class = null, string $format = null, array $context = array())
+    {
+        if (null === $class || !$this->metadataFactory->hasMetadataFor($class)) {
+            return $propertyName;
+        }
+
+        $classMetadata = $this->metadataFactory->getMetadataFor($class);
+
+        $attributesMetadata = $classMetadata->getAttributesMetadata();
+
+        foreach ($attributesMetadata as $name => $metadata) {
```

Comment:
Can you cache this in a private static property to avoid looping for every property?